lib/filterx/filterx-metrics.c

```diff
@@ -110,6 +145,17 @@ filterx_metrics_get_stats_counter(FilterXMetrics *self, StatsCounterItem **count
       return TRUE;
     }
 
+  /*
+   * We need to delay the optimization to the first get() call,
+   * as we don't have stats options when FilterXExprs are being created.
+   */
+  if (!self->is_optimized)
+    {
+      if (!_optimize(self))
```

Comment:
I think this needs to be protected by a lock, quite probably we can reuse the starts_lock() in this case.

Reply:
Nice catch! Fixed in a new commit.